src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -463,7 +483,8 @@ private static void MarkTransitiveOrigin(Dictionary<string, TransitiveEntry> tra
                     cachedEntry[fxRidEntry] = new List<PackageReference>();
                 }
 
-                if (!cachedEntry[fxRidEntry].Contains(top)) // Dictionary value is a List. If perf. is bad, change to HashSet.
+                // In order to hide Projects from the PM UI, we don't add them to the transitive origins cache
```

Comment:
Can you help me understand why the transitive origins cache is where the project refs are missing? I'm not sure how that translated to the transitive list of packages that exist. 

Another thing is that we might be mixing the model & view here, since this is really just data. If the view needs to do something different, maybe it needs to be handled there instead.

Reply:
I'm not sure if I understand the first question, maybe explaining why we check transitive origins answers you. When we construct the transitive dependencies list, we look at the assets file, and it contains projects and packages dependencies so at the end we have a list with all packages no matter who brough it. 

The problem here is that since we cannot correlate a package to an installed package, since they were brought by a project, we don't have a transitive origin and we don't display them (https://github.com/NuGet/NuGet.Client/pull/5524/files). This caused errors between packages list and vulnerabilities indicators, https://github.com/NuGet/Home/issues/13068. So, at the end transitive origins is the way we differentiate between packages brought by packages and projects.

About mixing model and view I agree that is not the best option, but I'm afraid that if we start using a new type to represent both packages and projects it will be a huge change. Right now, the way we determine transitive packages in the UI is by filtering out `PackageReferences` without transitive origins https://github.com/NuGet/NuGet.Client/pull/5524/files

Reply:
I think I get what it's supposed to say now. 

Why wouldn't we just try not to return the project in the list of installed packages? Are projects today included in the return type for the list? 
Note that this API is used for the get installed packages extensibility APIs as well, so we want to make sure we're not returning projects there.

Reply:
Projects are not being added to the returned installed packages list. I'm adding projects to a copy of the installed packages so I can calculate the transitive origin for their packages.

Reply:
This comment suggests that the transitive origins cache has a direct impact on what is shown in the UI. That should not be. I may be misremembering how it's used. I would love to understand this comment in more detail to make sure the comment is reflective of the code or that we see if there's an implicit effect we need to consider.